core/trino-parser/src/main/java/io/trino/sql/RowPatternFormatter.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql;
+
+import io.trino.sql.tree.AnchorPattern;
+import io.trino.sql.tree.AstVisitor;
+import io.trino.sql.tree.EmptyPattern;
+import io.trino.sql.tree.ExcludedPattern;
+import io.trino.sql.tree.GroupedPattern;
+import io.trino.sql.tree.Node;
+import io.trino.sql.tree.OneOrMoreQuantifier;
+import io.trino.sql.tree.PatternAlternation;
+import io.trino.sql.tree.PatternConcatenation;
+import io.trino.sql.tree.PatternPermutation;
+import io.trino.sql.tree.PatternVariable;
+import io.trino.sql.tree.QuantifiedPattern;
+import io.trino.sql.tree.RangeQuantifier;
+import io.trino.sql.tree.RowPattern;
+import io.trino.sql.tree.ZeroOrMoreQuantifier;
+import io.trino.sql.tree.ZeroOrOneQuantifier;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.joining;
+
+public final class RowPatternFormatter
+{
+    private RowPatternFormatter() {}
+
+    public static String formatPattern(RowPattern pattern)
+    {
+        return new Formatter().process(pattern, null);
+    }
+
+    public static class Formatter
+            extends AstVisitor<String, Void>
+    {
+        @Override
+        protected String visitNode(Node node, Void context)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected String visitRowPattern(RowPattern node, Void context)
+        {
+            throw new UnsupportedOperationException(format("not yet implemented: %s.visit%s", getClass().getName(), node.getClass().getSimpleName()));
+        }
+
+        @Override
+        protected String visitPatternAlternation(PatternAlternation node, Void context)
+        {
+            return node.getPatterns().stream()
+                    .map(child -> process(child, context))
+                    .collect(joining(" | "));
+        }
+
+        @Override
+        protected String visitPatternConcatenation(PatternConcatenation node, Void context)
+        {
+            return node.getPatterns().stream()
+                    .map(child -> process(child, context))
+                    .collect(joining(" "));
+        }
```

Comment:
This doesn't handle precedence properly. For instance:

```java
System.out.println(formatSql(
                new PatternConcatenation(
                        ImmutableList.of(
                                new PatternAlternation(ImmutableList.of(
                                        new PatternVariable(location(1, 1), identifier("a")),
                                        new PatternVariable(location(1, 1), identifier("b")))
                                ),
                                new PatternAlternation(ImmutableList.of(
                                        new PatternVariable(location(1, 1), identifier("c")),
                                        new PatternVariable(location(1, 1), identifier("d")))
                                )))));
```

produces:

```
a | b c | d
```

which then parses as 

```
a | (b c) | d
```